launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+   * - how to execute the command in general.
+   *
+   * The name is easy: PATH on Win32, DYLD_LIBRARY_PATH on MacOS, LD_LIBRARY_PATH elsewhere.
+   *
+   * On Unix-like, we're assuming bash is available. So we print one argument per line to
+   * the output, and use bash's array handling to execute the right thing.
+   *
+    * For Win32, see {@link #prepareForWindows(List<String>,String)}.
+   */
+  protected List<String> prepareForOs(List<String> cmd,
+      String libPath,
+      Map<String, String> env) {
+
+    // If SPARK_HOME does not come from the environment, explicitly set it
+    // in the child's environment.
+    Map<String, String> childEnv = env;
+    if (System.getenv("SPARK_HOME") == null && !env.containsKey("SPARK_HOME")) {
+      childEnv = new HashMap<String, String>(env);
+      childEnv.put("SPARK_HOME", sparkHome);
+    }
+
+    if (isWindows()) {
+      return prepareForWindows(cmd, libPath, childEnv);
+    }
+
+    if (isEmpty(libPath) && childEnv.isEmpty()) {
+      return cmd;
+    }
+
+    List<String> newCmd = new ArrayList<String>();
+    newCmd.add("env");
+
+    if (!isEmpty(libPath)) {
+      String envName = getLibPathEnvName();
+      String currEnvValue = getenv(envName);
+      String newEnvValue = join(File.pathSeparator, currEnvValue, libPath);
+      newCmd.add(String.format("%s=%s", envName, newEnvValue));
+    }
+    for (Map.Entry<String, String> e : childEnv.entrySet()) {
+      newCmd.add(String.format("%s=%s", e.getKey(), e.getValue()));
+    }
+    newCmd.addAll(cmd);
+    return newCmd;
+  }
+
+  protected String shQuote(String s) {
+    StringBuilder quoted = new StringBuilder();
+    boolean hasWhitespace = false;
+    for (int i = 0; i < s.length(); i++) {
+      if (Character.isWhitespace(s.codePointAt(i))) {
+        quoted.append('"');
+        hasWhitespace = true;
+        break;
+      }
+    }
+
+    for (int i = 0; i < s.length(); i++) {
+      int cp = s.codePointAt(i);
+      switch (cp) {
+        case '\'':
+          if (hasWhitespace) {
+            quoted.appendCodePoint(cp);
+            break;
+          }
+        case '"':
+        case '\\':
+          quoted.append('\\');
+          // Fall through.
+        default:
+          if (Character.isWhitespace(cp)) {
+            hasWhitespace=true;
+          }
+          quoted.appendCodePoint(cp);
+      }
+    }
+    if (hasWhitespace) {
+      quoted.append('"');
+    }
+    return quoted.toString();
+  }
+
+  // Visible for testing.
+  List<String> parseOptionString(String s) {
+    List<String> opts = new ArrayList<String>();
+    StringBuilder opt = new StringBuilder();
+    boolean inOpt = false;
+    boolean inSingleQuote = false;
+    boolean inDoubleQuote = false;
+    boolean escapeNext = false;
+    boolean hasData = false;
+
+    for (int i = 0; i < s.length(); i++) {
+      int c = s.codePointAt(i);
+      if (escapeNext) {
+        if (!inOpt) {
+          inOpt = true;
+        }
```

Comment:
actually I'm not sure why this is set here. Since this method seems general enough to be used beyond java opts, you might have an option that ends with a backslash (i.e. you might have `\\ ` two backslashes followed by a space, in which case we're not `inOpt` anymore right?)